### PR TITLE
Avoid recursion and allow building cyclic graphs

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Avoid recursion. Not sure we ever get to graphs of a size where this would matter, but still.
- Allow building graphs with cycles. Main motivation was to be able to visualize them, which could be valuable for debugging.

### CODE REVIEW

File: src/sciline/pipeline.py [hunk 1]
```diff
@@ -191,20 +190,22 @@ def build(self, tp: Type[T], /) -> Graph:
...
-            if name != 'return'
-        }
-        graph: Graph = {tp: (provider, args)}
-        try:
+        graph: Graph = {}
+        stack: List[Tuple[Type[T], Dict[TypeVar, Key]]] = [(tp, {})]
+        while stack:
+            tp, bound = stack.pop()
+            provider: Callable[..., T]
+            provider, bound = self._get_provider(tp)
```
Comment (YooSunYoung) [inline src/sciline/pipeline.py:198]: If `bound` is going to be retrieved by `self._get_provider`
`stack` doesn't need to remember `bound`, right...?

File: tests/pipeline_test.py [hunk 2]
```diff
@@ -492,16 +491,27 @@ class A(sl.Scope[T], int):
...
+def test_computing_graph_with_cycle_raises_CycleError() -> None:
+    def f(x: int) -> float:
+        return float(x)
+
+    def g(x: float) -> int:
+        return int(x)
+
+    pipeline = sl.Pipeline([f, g])
+    with pytest.raises(sl.scheduler.CycleError):
+        pipeline.compute(int)
```
Comment (YooSunYoung) [inline tests/pipeline_test.py:514]: Just a question,
Will it use `DaskScheduler` in `tox -e dask` and `NaiveScheduler` in `tox -e py38`...?

Comment (SimonHeybrock) [reply tests/pipeline_test.py:514]: Yes.

File: src/sciline/pipeline.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/sciline/pipeline.py:198]: Good point, fixed!

**Review by YooSunYoung — APPROVED** [review]

**Pull request merged**